test/test_binance_market.py

```diff
@@ -407,8 +407,8 @@ def delayed_time():
             time_offset = BinanceTime.get_instance().time_offset_ms
             print("offest", time_offset)
             # check if it is less than 5% off
-            self.assertTrue(time_offset > 0)
-            self.assertTrue(abs(time_offset - 30.0 * 1e3) < 1.5 * 1e3)
+            self.assertTrue(time_offset != 0)
```

Comment:
This is the wrong change to fix this test case.

The problem it didn't work wasn't that the checks were wrong, but the waiting times were wrong.

BinanceTime takes 60 seconds to poll for a new offset. The first `asyncio.sleep()` was pretty much unnecessary, because we haven't patched the delayed time function into the `binance_market` module yet. The second `asyncio.sleep()` afterwards was too short - 5 seconds wasn't enough time for the BinanceTime polling logic to calculate a new time offset.

The original checks were correct. What you should have changed is the waiting times instead. Also, the `time_offset` after a proper time offset update should be pretty large. So instead of checking for >0 you should actually check for something like >10000 (10 seconds).

Reply:
Two more things you need to be aware of here... are...

 1. Line 400 is not effective here... the BinanceTime update loop is already running and sleeping by the time it runs, so it's already delaying the next update by 60 seconds. What should have happened is the BinanceTime instance should have been stopped and then started again right after setting a shorter polling interval. There's another bug in BinanceTime here in which once it's stopped it cannot be started, because the stop function does not set the polling task to None. You should fix that as well.
 2. The offset time returned is taken as the median of multiple samples gathered. So it can happen that BinanceTime gathered a bunch of samples before you patched its `time.time()` function, and then you patch, and then it takes a few very large delta samples... and then it returns a median of <1s delta. So you need to be mindful of that as well.

So, if you want to fully fix this test case, and also make it fast s.t. you don't need to wait 1 minute+ for it to finish... you need to do the following:

 1. Fix BinanceTime.stop() - it needs to both set the old task object to None, and also clear out the old `_time_offset_ms` samples.
 2. Change line 400 to set the poll interval to 1 second. Even 3 seconds is too long.
 3. Remove the sleep at line 401.
 4. Stop and start the BinanceTime instance after that, so it'll start polling for offsets frequently.
 5. Change line 410 to check for `time_offset > 10000` rather than just 0.
 6. There's no need to sleep for 60 seconds+ anywhere in the test case.

Once you do that, you should see the intended result of having a large offset time calculated from BinanceTime.

Reply:
One more thing... after the test completed (whether it succeeded or failed), you shouldn't leave a frequently polling BinanceTime running for other test cases. So you should try.. finally.. the test case and restore the polling interval and restart BinanceTime again in the finally clause.

Reply:
@veecos2 feel free to hand this back to @michaelc9 if you feel like it's too much for you - since you aren't likely to be familiar with the Binance market connector and the details behind BinanceTime. However, if you'd like to take this up and learn a little bit deeper into the Binance market connector, I'm happy to explain the details to you when you're back in Tuesday.